### PR TITLE
@observe and @validate docstrings

### PR DESCRIPTION
**Pull request by SylvainCorlay**
- The `@validate` and `@observe` decorators docstrings were missing the documentation of the signature of the change and validation handlers.
- I added an important note in the docstring of the `@validate` decorator, which is a recommendation for authors of cross validator functions.

Let me explain in more details here.
- the owner is passed to the cross validator function, because we want the validation / coercion of the proposed value to be able to read other attributes of the owner, for example, to enforce min <= value <= max in a slider (flooring and capping proposed value with `min` and `max`). 
- as a consequence, the author of such cross validator could also potentially make changes to other attributes (pushing `min` and `max` around when changing the value).

The note is there to explain that one should **never** do the latter. The reason is that such changes may not commute and are applied in arbitrary orders when existing the `hold_trait_notification` context manager.

In short:
 - coercing the proposed value is **OK**

    (flooring and capping proposed value with `min` and `max`).
 - changing other attributes of the owner is **NOT OK**
 
   (pushing `min` and `max` around with the proposed value).

This would be easy to enforce if Python had a `const` keyword.

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
@ellisonbg @minrk

**Comment by minrk** [comment]
Sensible, thanks for the thorough docstring. Python being what it is, we can't (and shouldn't) *force* the validator functions to be well behaved, but a docstring indicating what makes sense is great.

**Comment by rmorshea** [comment]
How important is the rollback that occurs after a `TraitError`? That's why this isn't resolvable.

**Comment by SylvainCorlay** [comment]
This is not so much related with the rollback as it is that `modification1 + modification2 != modification2 + modification1`.
